container/container.go

```diff
@@ -1,105 +1,243 @@
+var statuses = []struct {
+	container bool
+	service   bool
+	status    Status
+}{
+	{service: true, container: true, status: RUNNING},
+	{service: true, container: false, status: STARTING},
+	{service: false, container: true, status: RUNNING}, // This is actually stopping
+	{service: false, container: false, status: STOPPED},
+}
+
+// buildResponse is the object that is returned by the docker build api.
+type buildResponse struct {
+	Stream string `json:"stream"`
+	Error  string `json:"error"`
+}
+
 // Container describes the API of container package.
 type Container interface {
 	Build(path string) (tag string, err error)
 	CreateNetwork(namespace string) (id string, err error)
 	DeleteNetwork(namespace string) error
-	FindContainer(namespace string) (types.ContainerJSON, error)
-	FindNetwork(namespace string) (types.NetworkResource, error)
-	FindService(namespace string) (swarm.Service, error)
-	ListServices(labels ...string) ([]swarm.Service, error)
-	ListTasks(namespace string) ([]swarm.Task, error)
-	Namespace(namespace string) string
-	SharedNetworkID() (networkID string, err error)
+	SharedNetworkID() string
 	StartService(options ServiceOptions) (serviceID string, err error)
-	Status(namespace string) (StatusType, error)
 	StopService(namespace string) (err error)
-	TasksError(namespace string) ([]string, error)
 	DeleteVolume(name string) error
 }
 
 // DockerContainer provides high level interactions with Docker API for MESG.
 type DockerContainer struct {
 	// client is a Docker client.
-	client docker.CommonAPIClient
-
-	// callTimeout is the timeout value for Docker API calls.
-	callTimeout time.Duration
-
-	// defaultStopGracePeriod is the timeout value between stopping a container and killing it.
-	defaultStopGracePeriod time.Duration
+	client client.CommonAPIClient
 
 	// namespace prefix.
 	nsprefix string
-}
 
-// Option is a configuration func for Container.
-type Option func(*DockerContainer)
+	// sharedNetworkID is cache for an id network.
+	sharedNetworkID string
+}
 
 // New creates a new Container with given options.
-func New(nsprefix string, options ...Option) (*DockerContainer, error) {
+func New(nsprefix string) (*DockerContainer, error) {
+	client, err := client.NewClientWithOpts(client.FromEnv, client.WithTimeout(clientHTTPTimeout))
+	if err != nil {
+		return nil, err
+	}
+	client.NegotiateAPIVersion(context.Background())
+
 	c := &DockerContainer{
-		callTimeout:            60 * time.Second,
-		defaultStopGracePeriod: 10 * time.Second, // default docker value
-		nsprefix:               nsprefix,
+		client:   client,
+		nsprefix: nsprefix,
+	}
+
+	if err := c.isSwarmInit(); err != nil {
+		return nil, err
+	}
+
+	if err := c.createSharedNetwork(); err != nil {
+		return nil, err
 	}
-	for _, option := range options {
-		option(c)
+	return c, nil
+}
+
+// Build builds a docker image.
+func (c *DockerContainer) Build(path string) (tag string, err error) {
+	excludeFiles, err := readDockerIgnoreFile(path)
```

Comment:
This suggestion could be done in another PR.

Reply:
The fix (when we run directly the binary on osx and not in docker) is the following
```go
archive.Untar(resp.Body, path, &archive.TarOptions{
	ChownOpts: &idtools.IDPair{
		UID: os.Geteuid(),
		GID: os.Getegid(),
	},
})
```

Reply:
ok so it's the untar that cause problem, not the tar ;)

Reply:
@krhubert why not following what docker cli is doing? you didn't answer anything. https://github.com/mesg-foundation/engine/pull/1400#discussion_r334261545

Reply:
So, I have already updated the code to use cli.

About that part. This should be done in cli (like we had in previous cli). Also if you run build in cli that part should be done, otherwise server modifies service somehow and I'm not sure if that is desirable behavior.